core/input/input.cpp

```diff
@@ -700,29 +711,39 @@ void Input::_parse_input_event_impl(const Ref<InputEvent> &p_event, bool p_is_em
 			continue;
 		}
 
-		Action &action = action_state[E.key];
+		int device_id = p_event->get_device();
+
+		ActionState &action_state = action_states[E.key];
+
 		if (!p_event->is_echo()) {
 			if (p_event->is_action_pressed(E.key)) {
-				if (!action.pressed) {
-					action.pressed_physics_frame = Engine::get_singleton()->get_physics_frames();
-					action.pressed_process_frame = Engine::get_singleton()->get_process_frames();
+				if (!action_state.cache.pressed) {
+					action_state.pressed_physics_frame = Engine::get_singleton()->get_physics_frames();
+					action_state.pressed_process_frame = Engine::get_singleton()->get_process_frames();
 				}
-				action.pressed |= ((uint64_t)1 << event_index);
-			} else {
-				action.pressed &= ~((uint64_t)1 << event_index);
-				action.pressed &= ~(1 << MAX_EVENT); // Always release the event from action_press() method.
+				ActionState::DeviceState &device_state = action_state.device_states[device_id];
+				device_state.pressed[event_index] = true;
+				device_state.strength[event_index] = p_event->get_action_strength(E.key);
+				device_state.raw_strength[event_index] = p_event->get_action_raw_strength(E.key);
 
-				if (!action.pressed) {
-					action.released_physics_frame = Engine::get_singleton()->get_physics_frames();
-					action.released_process_frame = Engine::get_singleton()->get_process_frames();
+				action_state.exact = InputMap::get_singleton()->event_is_action(p_event, E.key, true);
+				_update_action_cache(E.key, action_state);
+			} else {
+				ActionState::DeviceState &device_state = action_state.device_states[device_id];
+				device_state.pressed[event_index] = false;
+				device_state.strength[event_index] = 0.0;
+				device_state.raw_strength[event_index] = 0.0;
```

Comment:
You should still assign strength for released events (at least the raw one, not sure about *strength*). It's relevant for joypad axes.